### PR TITLE
feat: source install variant, better default stages

### PR DESCRIPTION
**Pull request by scop**
See individual commits for details.

This could have been split to three different PR's, as the commits aren't that closely related, however they would cause conflicts and they do build on each other.

The bot that updates the mirror might need a tweak to get the newly added dependency version updated, too.

### CODE REVIEW

File: .pre-commit-hooks.yaml [hunk 1]
```diff
@@ -1,10 +1,16 @@
...
-    description: 'A pluggable and configurable code formatting platform written in Rust.'
+    description: 'Pluggable and configurable code formatting platform (npm install)'
     entry: dprint fmt
     language: node
     'types': [text]
-    args: []
-    require_serial: false
     additional_dependencies: ["dprint@0.32.2"]
-    minimum_pre_commit_version: '0'
+    stages: [commit, merge-commit, push, manual]
```
Comment (adamchainz) [inline .pre-commit-hooks.yaml:8]: I don't think the hook should run on multiple stages like this - the expectation is generally the default stages only, users can customize as needed

**Review by adamchainz — COMMENTED** [review]
I don’t really see the value in avoiding use of node. pre-commit installs node for you, with nodeenv, and then a binary of dprint can be used. Using a source install requires cargo to be set up, and a potentially long compile step.

Also as you noted, pre-commit-mirror-maker will undo all the changes you’ve made here. If you don’t update the tool to support multiple hooks and the other options, I’m not sure who will.

ANyway, I would say if you want a source install, use your own hook definition. It’s not much code.

**Comment by adamchainz** [comment]
Ah I see Anthony’s comment on stages:

> code formatters should set `[commit, merge-commit, push, manual]` as the reasonable default. 

I’m not sure how much he actually believes this comment. The major code formatters I’ve used ([black](https://github.com/psf/black/blob/main/.pre-commit-hooks.yaml), [prettier](https://github.com/pre-commit/mirrors-prettier/blob/main/.pre-commit-hooks.yaml)) use just the default stages.

**Comment by scop** [comment]
Default stages (=all) tend to work fine only because use of anything besides commit is quite rare. As soon as one adds some other stages, the redundancy is immediately visible.

**Comment by scop** [comment]
The problem with node here is that it's really unrelated to the task at hand. It's a complete runtime/language required just for the sake of installing a tool, which to me is excessive, no matter if it's automatic or not.

Installing a prebuilt binary of dprint is also not something that is appreciated in the first place in many setups, no matter if it's pulled in by this installer or by some other means. I'm sure you're aware that choices to install it via pre-commit are quite limited, and having pre-commit install the tool is a major chunk of pre-commit's usefulness in the first place.

The long compile time will happen, sure, but the source install hook is not the "default", it doesn't happen except when the cache needs a refresh, and I don't think it should come as a surprise for people who explicitly opt in to the source variant.

**Comment by adamchainz** [comment]
Okay. I’m not using dprint at the moment. Would you like to take over this repo?

**Comment by scop** [comment]
Not at this point, thanks -- I'm just preparing to evaluate dprint myself.

**Pull request closed without merging**